modules/invibesBidAdapter.js

```diff
@@ -0,0 +1,438 @@
+      invibes.ivLogger.info('Bid auction started at ' + bidModel.AuctionStartTime + ' . Invibes registered the bid at ' + now + ' ; bid request took a total of ' + (now - bidModel.AuctionStartTime) + ' ms.');
+    } else {
+      utils.logInfo('Invibes Adapter - Incorrect Placement Id: ' + bidRequest.params.placementId);
+    }
+  }
+
+  return bidResponses;
+}
+
+function generateRandomId() {
+  return (Math.round(Math.random() * 1e12)).toString(36).substring(0, 10);
+}
+
+function getDocumentLocation(topWin) {
+  return topWin.location.href.substring(0, 300).split(/[?#]/)[0];
+}
+
+function parseQueryStringParams() {
+  var params = {};
+  try { params = JSON.parse(localStorage.ivbs); } catch (e) { }
+  var re = /[\\?&]([^=]+)=([^\\?&#]+)/g;
+  var m;
+  while ((m = re.exec(window.location.href)) != null) {
+    if (m.index === re.lastIndex) {
+      re.lastIndex++;
+    }
+    params[m[1].toLowerCase()] = m[2];
+  }
+  return params;
+}
+
+function getBiggerSize(array) {
+  var result = [0, 0];
+  for (var i = 0; i < array.length; i++) {
+    if (array[i][0] * array[i][1] > result[0] * result[1]) {
+      result = array[i];
+    }
+  }
+  return result;
+}
+
+function getTopMostWindow() {
+  var res = window;
+
+  try {
+    while (top !== res) {
+      if (res.parent.location.href.length) { res = res.parent; }
+    }
+  } catch (e) { }
+
+  return res;
+}
+
+function renderCreative(bidModel) {
+  return `<html>
+        <head><script type='text/javascript'>inDapIF=true;</script></head>
+          <body style='margin : 0; padding: 0;'>
+          creativeHtml
+          </body>
+        </html>`
+    .replace('creativeHtml', bidModel.CreativeHtml);
+}
+
+function getCappedCampaignsAsString() {
+  var key = 'ivvcap';
+
+  var loadData = function () {
+    return JSON.parse(localStorage.getItem(key)) || {};
+  };
+
+  var saveData = function (data) {
+    localStorage.setItem(key, JSON.stringify(data));
+  };
+
+  var clearExpired = function () {
+    var now = new Date().getTime();
+    var data = loadData();
+    var dirty = false;
+    Object.keys(data).forEach(function (k) {
+      var exp = data[k][1];
+      if (exp <= now) {
+        delete data[k];
+        dirty = true;
+      }
+    });
+    if (dirty) {
+      saveData(data);
+    }
+  };
+
+  var getCappedCampaigns = function () {
+    clearExpired();
+    var data = loadData();
+    return Object.keys(data)
+      .filter(function (k) { return data.hasOwnProperty(k); })
+      .sort()
+      .map(function (k) { return [k, data[k][0]]; });
+  };
+
+  return getCappedCampaigns()
+    .map(function (record) { return record.join('='); })
+    .join(',');
+}
+
+function initLogger() {
+  const noop = function () { };
+
+  if (localStorage && localStorage.InvibesDEBUG) {
+    return window.console;
+  }
+
+  return { info: noop, error: noop, log: noop, warn: noop, debug: noop };
+}
+
+/// Local domain cookie management =====================
+var Uid = {
```

Comment:
Can you share how are you using these cookie functions and what is their purpose ?

Reply:
ivNoCookie -> no cookies should be set for the user
ivbsdid -> a user id cookie. The entire 'Uid' +'initDomainId' is logic is for filtering (the user must return at least several times in multiple days. Else it is considered a temporary id and not sent with the bid request)
IvbsCampIdsLocal -> list of clicked campaign ids of the user
random test cookie -> to check for top cookie domain. We remove it afterwards